astropy/io/ascii/tests/test_write.py

```diff
@@ -534,3 +535,46 @@ def test_byte_string_output(fast_writer):
     out = StringIO()
     ascii.write(t, out, fast_writer=fast_writer)
     assert out.getvalue().splitlines() == ['col0', 'Hello', 'World']
+
+
+@pytest.mark.parametrize('names, include_names, exclude_names, formats, issues_warning', [
+    (['x', 'y'], ['x', 'y'], ['x'], {}, True),
+    (['x', 'y'], ['x', 'y'], ['y'], {}, True),
+    (['x', 'y'], ['x', 'y'], [], {'p':'%d', 'q':'%f'}, True),
+    (['x', 'y'], ['x', 'y'], [], {'z':'%f'}, True),
+    (['x', 'y'], ['x', 'y'], [], {'x':'%d'}, False),
+    (['x', 'y'], ['x', 'y'], [], {'p':'%d', 'y':'%f'}, True),
+    (['x', 'y'], ['x', 'y'], [], {}, False)
+])
+def test_names_with_formats(names, include_names, exclude_names, formats, issues_warning):
+    """Test the fix for #4508 where i set the test cases in base of issues_warning
+    if issues_warning value is true means warning accure otherwise issues_warning
+    value false.Check that columns of include_names are same with colomns of
+    exclude_names and Check that columns  in formats specifier exist in the
+    issues_warning table when writing.
+    """
+    t = table.Table([[1,2,3],[4.1,5.2,6.3]])
+    with catch_warnings(AstropyWarning) as ASwarn:
+        out = StringIO()
+        ascii.write(t, out, names=names, include_names=include_names, \
+        exclude_names=exclude_names, formats=formats)
+    assert (issues_warning == (len(ASwarn)>0))
```

Comment:
This supposes that multiple warnings can be raised, is it the case ? If yes, the test could pass with a warning which is not the one you expect, so it would require a more specific test. But here I guess you can test that  `len(ASwarn) == 1`, or put the number of expected warnings in the fixture.